internal/providers/virtualbox/virtualbox.go

```diff
@@ -12,40 +12,66 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// The virtualbox provider fetches the configuration from raw data on a partition
-// with the GUID 99570a8a-f826-4eb0-ba4e-9dd72d55ea13
+// The virtualbox provider fetches the configuration from the /Ignition/Config
+// guest property.
 
 package virtualbox
 
+// We want at least this warning, since the default C behavior of
+// assuming int foo(int) is totally broken.
+
+// #cgo CFLAGS: -Werror=implicit-function-declaration
+// #include <linux/vbox_err.h>
+// #include <stdlib.h>
+// #include "virtualbox.h"
+import "C"
+
 import (
 	"bytes"
-	"io/ioutil"
-	"os"
-	"path/filepath"
+	"fmt"
+	"unsafe"
 
 	"github.com/coreos/ignition/v2/config/shared/errors"
 	"github.com/coreos/ignition/v2/config/v3_4_experimental/types"
-	"github.com/coreos/ignition/v2/internal/distro"
 	"github.com/coreos/ignition/v2/internal/providers/util"
 	"github.com/coreos/ignition/v2/internal/resource"
 
 	"github.com/coreos/vcontext/report"
 )
 
 const (
-	partUUID = "99570a8a-f826-4eb0-ba4e-9dd72d55ea13"
+	guestProperty = "/Ignition/Config"
 )
 
 func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
-	f.Logger.Debug("Attempting to read config drive")
-	rawConfig, err := ioutil.ReadFile(filepath.Join(distro.DiskByPartUUIDDir(), partUUID))
-	if os.IsNotExist(err) {
-		f.Logger.Info("Path to ignition config does not exist, assuming no config")
-		return types.Config{}, report.Report{}, errors.ErrEmpty
-	} else if err != nil {
-		f.Logger.Err("Error reading ignition config: %v", err)
+	f.Logger.Debug("reading Ignition config from VirtualBox guest property")
+	rawConfig, err := fetchProperty(guestProperty)
+	if err != nil {
+		if err == errors.ErrEmpty {
+			f.Logger.Info("VirtualBox guest property %q does not exist; assuming no config", guestProperty)
+		}
 		return types.Config{}, report.Report{}, err
 	}
 	trimmedConfig := bytes.TrimRight(rawConfig, "\x00")
 	return util.ParseConfig(f.Logger, trimmedConfig)
 }
+
+func fetchProperty(name string) ([]byte, error) {
+	cName := C.CString(name)
+	defer C.free(unsafe.Pointer(cName))
+	buf := unsafe.Pointer(nil)
+	defer C.free(buf)
+	var size C.size_t
+
+	ret, errno := C.virtualbox_get_guest_property(cName, &buf, &size)
+	if ret == C.VERR_GENERAL_FAILURE && errno != nil {
```

Comment:
I think this error handling is correct, but also a bit sketchy to read.
I would suggest inverting it so that there is an outer `if ret != C.VINF_SUCCESS`, and then inner error formatters based on `ret` and `errno` as you like.